src/Microsoft.Android.Build.BaseTasks/MSBuildExtensions.cs

```diff
@@ -252,34 +252,69 @@ public static void SetDestinationSubPath (this ITaskItem assembly)
 		/// <summary>
 		/// IBuildEngine4.RegisterTaskObject, but adds the current assembly path into the key
 		/// </summary>
+		[Obsolete ("Use RegisterTaskObjectAssemblyLocal (engine, key, value, allowEarlyCollection, lifetime, projectSpecific) instead.")]
 		public static void RegisterTaskObjectAssemblyLocal (this IBuildEngine4 engine, object key, object value, RegisteredTaskObjectLifetime lifetime, bool allowEarlyCollection = false) =>
-			engine.RegisterTaskObject ((AssemblyLocation, key), value, lifetime, allowEarlyCollection);
+			RegisterTaskObjectAssemblyLocal (engine, key, value, lifetime, allowEarlyCollection: false, projectSpecific: true);
+
+		/// <summary>
+		/// IBuildEngine4.RegisterTaskObject, but adds the current assembly path into the key
+		/// </summary>
+		public static void RegisterTaskObjectAssemblyLocal (this IBuildEngine4 engine, object key, object value, RegisteredTaskObjectLifetime lifetime, bool allowEarlyCollection = false, bool projectSpecific = true) =>
```

Comment:
Public API-wise, I'm not certain that we want `bool projectSpecific` to be the overload.  I'm torn between different ideas:

 1. An enum-based parameter, or
 2. Generics(?)!

The benefit to an enum-based approach is that we can extend it to include additional data over time:

```csharp
[Flags]
public enum RegisterTaskObjectKey {
    None = 0,
    IncludeProjectFile = 1 << 0,
    // Can add other values in the future…
}
```

This would allow `engine.RegisterTaskObject("key1", value, lifetime, keys: RegisterTaskObjectKey.IncludeProjectFile)`, and if we need to include other data in the future, that can be reasonably done.

Then there is the "Generics!" idea:

```csharp
public static void RegisterTaskObjectAssemblyLocal<TRest>(
    this IBuildEngine4 engine,
    object key,
    object value,
    RegisteredTaskObjectLifetime lifetime,
    bool allowEarlyCollection = false,
    TRest? rest = default
);
```

where `rest` is part of the "tuple key", a'la `(AssemblyLocation, key, rest)`.  This would allow callers to include arbitrary data in the key for uniqueness, but might also be a bridge too far:

```
engine.RegisterTaskObject("key1", value, lifetime, rest: engine.ProjectFileOfTaskNode);
```

The other problem is that the 'Generics!' approach likely make it easier to be inconsistent, so maybe that isn't a good idea.

Reply:
I like the `[Flags]` `enum` idea. Can `lifetime` and `allowEarlyCollection` also be some of the flags?

Reply:
I went with the `[Flags]` .